### PR TITLE
[Experiment] [Novo] Check performance when extracting stuff from old shell 

### PR DESCRIPTION
**Pull request by damassi**
This PR is a basic working example of the new network of apps extracted from the main app shell. No changes have been made to our code, other than refactoring the express app to support both force and novo. 

TLDR; the lighthouse scores are improved slightly (see results below) however the TTI has been decreased significantly. There is still a lot of additional effort required in our journey to a performant front end.

### Core changes made

- Creates a minimal html page to mount novo assets.
- Creates a separate webpack build to generate decoupled assets for novo.
- Updates the current express server to host both force and novo.
- When a user visted a url, we served a fresh page template of raw html and assets, and skipped the jade pipeline completely

### package.json changes

The `yarn build*` variations are used to build production assets, with the exception of `yarn build:dev` which is being left for backwards compatibility.

* `yarn build`.- Builds all assets, however it no longer automatically cleans.
* `yarn build:assets` - Builds force assets only.
* `yarn build:assets:novo` - Builds novo assets only.
* `yarn build:server` - Build server code only.

The `yarn bundle-report*` variations are used to analyze final asset sizes.

* `yarn bundle-report:assets` - Displays the asset sizes for force.
* `yarn bundle-report:assets:novo` - Displays the assets sizes for novo.
* `yarn bundle-report:server` - Displays the asset sizes for the server.

The `yarn dev*` variations are used to run the application in development mode.

* `yarn dev` - Runs all dev commands in parallel.
* `yarn dev:relay` - Runs the relay compiler and if `watchman` is install it should automatically recompile changes.
* `yarn dev:server` - Runs the server in development mode.

### Running

To run in development mode:

```sh
$ yarn dev
```

To run in production mode:

```sh
yarn build && yarn start:prod:local
```

### Performance Insights

The current force bundle size totals 3.99 MiB, with ~2.15 MiB of common assets loading for every v2 page.

artsy-force-bundle

The current novo bundle size totals 1.22 MiB, with ~1 MiB of common assets loading for every v2 page.

artsy-novo-bundle

The Calibre results for the review app are showing significant improvements in TTI

calibre-names

### CODE REVIEW

File: webpack/envs/clientCommonConfig.js [hunk 1]
```diff
@@ -11,7 +11,13 @@ export const clientCommonConfig = {
   mode: env.nodeEnv,
   devtool: "source-map",
   stats: "normal", // or, `errors-only`
-  entry: getEntrypoints(),
+  entry: {
+    "artsy-v3": [
```
Comment (damassi) [inline webpack/envs/clientCommonConfig.js:15]: Only compiling the new app ATM to speed up compilation time

File: webpack/envs/serverConfig.js [hunk 2]
```diff
@@ -15,7 +15,8 @@ export const serverConfig = {
   node: {
     __dirname: true,
   },
-  entry: path.join(basePath, "src/index.js"),
+  // entry: path.join(basePath, "src/index.js"),
+  entry: path.join(basePath, "src/v3/index.js"),
```
Comment (damassi) [inline webpack/envs/serverConfig.js:19]: Only compiling the new server file, excluding everything else

**Comment by damassi** [comment]
Here are some initial scores: https://artsy.slack.com/archives/CE7HUL6AE/p1602356974049900

**Comment by eessex** [comment]
Just added a new commit that represents WIP towards a second deployment. We have duplicated the webpack config and added a new docker stage. @icirellik to follow up with notes on the mob session.

File: hokusai/staging.yml [hunk 3]
```diff
@@ -118,6 +118,10 @@ spec:
     - host: staging.artsy.net
       http:
         paths:
+          - path: /novo/
+            backend:
+              serviceName: force-web-novo-internal
+              servicePort: force-http
```
Comment (zephraph) [inline hokusai/staging.yml:124]: I'm not sure if it's intended for this to happen somewhere else, but the `force-web-novo-internal` service needs to be defined (usually above in the services section)

File: package.json [hunk 4]
```diff
@@ -10,11 +10,10 @@
   "scripts": {
     "acceptance-record": "scripts/acceptance_record.sh",
     "acceptance": "scripts/acceptance.sh",
-    "assets": "NODE_ENV=production yarn build",
-    "build:server": "BUILD_SERVER=true NODE_ENV=production yarn webpack",
-    "build": "yarn clean && yarn relay && BUILD_CLIENT=true yarn webpack",
-    "bundle-report": "NODE_ENV=production WEBPACK_ANALYZE=true yarn build",
-    "bundle-report:dev": "NODE_ENV=development WEBPACK_ANALYZE=true yarn build",
+    "assets": "NODE_ENV=production yarn webpack",
+    "build:dev": "yarn clean && yarn relay && BUILD_CLIENT=true yarn webpack",
```
Comment (damassi) [inline package.json:14]: How do we build server files now?

File: scripts/smoke_test.sh [hunk 5]
```diff
@@ -11,7 +11,7 @@ fi
 # # prepare a production build
 if [ ! -f server.dist.js ]; then
     yarn assets
-    yarn build:server
+    BUILD_SERVER=true yarn assets
```
Comment (damassi) [inline scripts/smoke_test.sh:14]: can we add this back to `package.json`? Its used pretty commonly

**Comment by icirellik** [comment]
Below is the loosely defined list of tasks I gathered from when we added the new builds:

* Parallelize novo builds in docker Buildkit.
* Cleanup novo Webpack configurations. Only include required plugins, loaders, etc.
* Cleanup docker novo build. Only include novo assets.
* Add novo build steps to the build pipeline.
* Verify that novo base path in Webpack configurations is working correctly and assets are being served from the proper location.
* Cleanup task names in package.json. Including adding additional novo build tasks.

File: hokusai/production.yml [hunk 6]
```diff
@@ -103,6 +106,119 @@ spec:
...
+        component: web
+      name: force-web-novo
+      namespace: default
+    spec:
+      containers:
+        - name: force-web-novo
+          env:
+            - name: PORT
+              value: "5000"
+            - name: DD_TRACE_AGENT_HOSTNAME
```
Comment (artsyjian) [inline hokusai/production.yml:137]: @eessex - Do we want to separate canonical and novo stats on Datadog?

File: Dockerfile [hunk 7]
```diff
@@ -195,25 +212,50 @@ RUN chown deploy:deploy $(pwd)
...
+COPY --chown=deploy:deploy --from=builder /app/public ./public
+COPY --chown=deploy:deploy --from=builder /app/src ./src
+
+# Server assets
+COPY --chown=deploy:deploy --from=builder /app/server.dist.js .
+COPY --chown=deploy:deploy --from=builder /app/server.dist.js.map .
 
 ENTRYPOINT ["/usr/bin/dumb-init", "--"]
-CMD ["yarn", "start"]
+CMD ["node", "--max_old_space_size=3072", "-r", "dotenv/config", "./server.dist.js"]
```
Comment (artsyjian) [inline Dockerfile:261]: @eessex - The changes we've made to `staging.yml/production.yml` create a 2nd set of pods, for novo. The pods will run the same Docker image (as the canonical pods). Shall they run the same `CMD`? Or, should we over-write it (can be done in k8s yml)? Or, were we planning to use the same `CMD` and toggle app-selection by ENV var, in that case, we can set the var (also can be done in k8s yml).

File: hokusai/production.yml [hunk 8]
```diff
@@ -103,6 +106,119 @@ spec:
...
+      maxSurge: 2
+      maxUnavailable: 0
+    type: RollingUpdate
+  template:
+    metadata:
+      labels:
+        app: force-novo
+        layer: application
+        component: web
+      name: force-web-novo
```
Comment (joeyAghion) [inline hokusai/production.yml:129]: Should maybe be `...-novo-web` so the prefixes are all consistently `force-novo-`.

Comment (artsyjian) [reply hokusai/production.yml:129]: Switching novo/web around should be fine. I'll update.

File: src/lib/middleware/assetMiddleware.ts [hunk 9]
```diff
@@ -18,6 +18,7 @@ export function assetMiddleware() {
     return (_req, res, next) => {
       res.locals.asset = filename => {
         let manifestFile = manifest[filename] || filename
+        // TODO: Is this path still used? If no remove it.
```
Comment (damassi) [inline src/lib/middleware/assetMiddleware.ts:21]: Referring to CDN_URL?

File: src/desktop/apps/artsy-v2/server.tsx [hunk 10]
```diff
@@ -32,6 +32,8 @@ const allRoutes = flatten(
   })
 )
 
+console.log(allRoutes)
```
Comment (joeyAghion) [inline src/desktop/apps/artsy-v2/server.tsx:35]: 👀

File: src/lib/middleware/assetMiddleware.ts (hunk 9 above)
Comment (joeyAghion) [reply src/lib/middleware/assetMiddleware.ts:21]: I think this was referring to whether the Force app receives any asset requests in production at all these days. If the references all point to s3 via the CDN, it shouldn't.

**Review by joeyAghion — APPROVED** [review]

File: src/lib/middleware/assetMiddleware.ts (hunk 9 above)
Comment (damassi) [reply src/lib/middleware/assetMiddleware.ts:21]: When building to prod we do looking with this CDN_URL (so this is used), but when building locally during dev force will receive requests.

**Review by izakp — APPROVED** [review]

File: src/common-app.js [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+app.use(logger)
+
+// Install common middlewares
+commonMiddlewareSetup(app)
+
+// novo must be first, the current app contains a catch all route.
+import novo from "./novo/index"
+app.use("/", novo)
+
+import current from "./current"
```
Comment (damassi) [inline src/common-app.js:23]: Do these `import`s get hoisted during babel transpilation?

File: src/common-middleware.js [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+
+import compression from "compression"
+import express from "express"
+import favicon from "serve-favicon"
+import glob from "glob"
+import path from "path"
+import sharify from "sharify"
+import siteAssociation from "artsy-eigen-web-association"
+
+function staticAssetMiddlewares(app) {
```
Comment (damassi) [inline src/common-middleware.js:13]: Since `express` is a singleton is it necessary to pass it as an argument?

File: src/index.js [hunk 13]
```diff
@@ -1,12 +1,38 @@
-const express = require("express")
+/* eslint-disable no-console */
+// TODO: Do we still need this.
+import 'source-map-support/register'
+
+// Do we still use regenerator?
+import "regenerator-runtime/runtime"
```
Comment (damassi) [inline src/index.js:6]: I'm about 90% sure this is safe to remove, and was only necessary for https://github.com/relay-tools/react-relay-network-modern#what-if-regeneratorruntime-is-not-defined, which i'm also about 90% sure has been fixed... We should try to ditch it and click around the new pages a bit.

File: src/index.js [hunk 14]
```diff
@@ -1,12 +1,38 @@
...
+
+// This must come before any other instrumented module.
+// See https://docs.datadoghq.com/tracing/languages/nodejs/ for more info.
+import "./lib/datadog"
+
+// TODO: Do we need JSDOM server-side?
+import "./lib/DOMParser"
```
Comment (damassi) [inline src/index.js]: @eessex - I believe this was related to some publishing pop up thing that was only relevant on the client?

File: src/lib/middleware/assetMiddleware.ts [hunk 15]
```diff
@@ -4,7 +4,7 @@ import chalk from "chalk"
 
 const { CDN_URL, NODE_ENV } = process.env
 
-export function assetMiddleware(manifestFileName = "manifest.json") {
+export default function assetMiddleware(manifestFileName = "manifest.json") {
```
Comment (damassi) [inline src/lib/middleware/assetMiddleware.ts:7]: Why `default`?

File: src/desktop/apps/authentication/helpers.setCookies.ts [hunk 16]
```diff
@@ -0,0 +1,15 @@
+import Cookies from "cookies-js"
+
+export const setCookies = options => {
```
Comment (eessex) [inline src/desktop/apps/authentication/helpers.setCookies.ts:3]: Why did this need to be moved out of the helpers file? This is a non-standard naming convention for Force.

File: src/common-app.js (hunk 11 above)
Comment (icirellik) [reply src/common-app.js:62]: It wouldn't make a difference either way, I'm mainly organizing this file for future refactoring.

File: src/lib/middleware/assetMiddleware.ts (hunk 15 above)
Comment (icirellik) [reply src/lib/middleware/assetMiddleware.ts:7]: I plan of normalizing all the current middlewares to follow the same pattern at which point a long term decision can be made. I would be hesitant to avoid language features because they are "difficult" to use and instead suggest we rely of tools such as linters and IDEs.

File: src/common-middleware.js (hunk 12 above)
Comment (icirellik) [reply src/common-middleware.js:56]: What do you mean by singleton in this case? Executing `express()` creates you a new object. Regardless, a second refactoring would clean this up.

File: src/novo/src/client.tsx [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+import ReactDOM from "react-dom"
+import { buildClientApp } from "v2/Artsy/Router/client"
+import { getAppNovoRoutes } from "v2/Apps/getAppNovoRoutes"
+import { data as sd } from "sharify"
+import { loadableReady } from "@loadable/component"
+
+import { mediator } from "lib/mediator"
+
+buildClientApp({
+  routes: getAppNovoRoutes(),
```
Comment (eessex) [inline src/novo/src/client.tsx:13]: Args should be just `routes`, see: https://github.com/artsy/force/blob/f431126f291c602c9f14354d16f2a333348fc28b/src/desktop/apps/artsy-v2/client.tsx#L12

`mediator` and `user` are now required by default here: https://github.com/artsy/force/blob/c1752c4c53d7d37424f4bea75194fb577678ec48/src/desktop/lib/buildClientAppContext.ts#L12:14

File: src/novo/src/server.ts [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+      const {
+        status,
+        styleTags,
+        scripts,
+        redirect,
+        bodyHTML,
+        headTags,
+      } = await buildServerApp(
+        {
+          req,
```
Comment (eessex) [inline src/novo/src/server.ts:49]: Server app context building has changed: https://github.com/artsy/force/blob/f431126f291c602c9f14354d16f2a333348fc28b/src/desktop/apps/artsy-v2/server.tsx#L72

File: src/novo/src/client.tsx (hunk 17 above)
Comment (eessex) [reply src/novo/src/client.tsx:13]: Relevant PRs for changes are:
https://github.com/artsy/force/pull/6482
https://github.com/artsy/force/pull/6511

File: src/index.js (hunk 14 above)
Comment (icirellik) [reply src/index.js]: I updated the comment when I discovered what it was used for.

`// Required for @artsy/react-html-parser which is included by @artsy/reaction`

File: src/desktop/apps/authentication/helpers.setCookies.ts (hunk 16 above)
Comment (icirellik) [reply src/desktop/apps/authentication/helpers.setCookies.ts:3]: It was removed while you were cleaning up the logged out user, this is the only function being used from the helpers file in novo.

File: src/lib/middleware/assetMiddleware.ts (hunk 15 above)
Comment (damassi) [reply src/lib/middleware/assetMiddleware.ts:7]: (blocking) For all new files being created we use named exports, per long-term artsy convention. As we discussed in slack, I'd encourage you to open an RFC for changing this pattern.

**Review by damassi — COMMENTED** [review]
@icirellik  - One blocking comment before merge: https://github.com/artsy/force/pull/6456#discussion_r520057414

File: src/lib/manifest.ts [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+import fs from "fs"
+import chalk from "chalk"
+
+const { CDN_URL, NODE_ENV } = process.env
+
+class ManifestManager {
+  private loadError = false
+  private manifest = {}
+
+  constructor(private manifestFileName = "manifest.json") {
```
Comment (damassi) [inline src/lib/manifest.ts:11]: How does that `private manifestFileName` arg work? Haven't seen this before

File: webpack/envs/clientNovoDevelopmentConfig.js [hunk 20]
```diff
@@ -0,0 +1,225 @@
...
+export const clientNovoDevelopmentConfig = {
+  stats: "normal",
+  parallelism: 100,
+  mode: env.webpackDebug ? "development" : env.nodeEnv,
+  devtool: "source-map",
+  entry: {
+    "artsy-novo": [
+      path.resolve(process.cwd(), "src/novo/src/client.tsx"),
+    ],
+  },
```
Comment (damassi) [inline webpack/envs/clientNovoDevelopmentConfig.js:20]: Do we need to duplicate all of this config? We should be able to use webpack smart merging to only replace the keys that change:

```
clientNovoDevelopmentConfig = {
   entry: {
     "artsy-novo": [
        path.resolve(process.cwd(), "src/novo/src/client.tsx"),
     ],
  },
}
```
Is (mostly) all that we'd need; and then the diff can be what we change here to help optimize bundle splitting. Same applies to the server config.

File: webpack/envs/clientNovoProductionConfig.js [hunk 21]
```diff
@@ -0,0 +1,235 @@
...
+const LoadablePlugin = require("@loadable/webpack-plugin")
+
+export const clientNovoProductionConfig = {
+  stats: "normal",
+  parallelism: 100,
+  mode: env.webpackDebug ? "development" : env.nodeEnv,
+  devtool: "source-map",
+  entry: {
+    "artsy-novo": [path.resolve(process.cwd(), "src/novo/src/client.tsx")],
+  },
```
Comment (damassi) [inline webpack/envs/clientNovoProductionConfig.js:20]: Same here, we can use smart merging to "extend" base artsy production config and update the `entry` key to point to new endpoint; no need to duplicate everything.

File: webpack/index.js [hunk 22]
```diff
@@ -36,11 +39,31 @@ const getClientConfig = () => {
   }
 }
 
+const getNovoClientConfig = () => {
+  switch (true) {
+    case env.isDevelopment:
+      console.log("[Force Novo] Building client-side development code...")
+      return clientNovoDevelopmentConfig
```
Comment (damassi) [inline webpack/index.js:46]: Something like: 
```jsx
return merge.smart(clientNovoDevelopmentConfig, clientCommonConfig)
```

File: scripts/acceptance.sh [hunk 23]
```diff
@@ -2,14 +2,13 @@
...
+export NODE_ENV=production
 
-  # CSS is only compiled during a development build?
-  export NODE_ENV=production
+if [ ! -f server.dist.js ]; then
+  yarn build:assets
+  yarn build:assets:novo
 
-  yarn assets
+  # TODO: We don't need the server to be built here, these tests only rely on assets.
```
Comment (eessex) [inline scripts/acceptance.sh:11]: Is this comment still accurate? Not sure why we build server when they rely on assets.

File: package.json [hunk 24]
```diff
@@ -25,22 +27,25 @@
...
+    "publish-assets:local": "yarn clean && yarn build:assets && yarn publish-assets",
     "relay": "relay-compiler",
     "sentry": "scripts/sentry.sh",
-    "start:dev": "NODE_ENV=development BUILD_CLIENT=true node --inspect --max_old_space_size=3072 -r dotenv/config ./src",
-    "start:prod:local": "NODE_ENV=production node --inspect --max_old_space_size=3072 -r dotenv/config ./server.dist.js",
-    "start:prod": "yarn assets && yarn build:server && NODE_ENV=production yarn start",
     "start": "scripts/start.sh",
+    "start:dev": "concurrently 'yarn start:dev:relay' 'yarn start:dev:server'",
+    "start:dev:server": "NODE_ENV=development BUILD_CLIENT=true node --inspect --max_old_space_size=3072 -r dotenv/config ./src/index.dev.js",
+    "start:dev:relay": "NODE_ENV=development yarn relay --watch",
```
Comment (damassi) [inline package.json:37]: We already have a `yarn relay` command, and yarn start runs `yarn relay --watch`. Whats the thinking behind this addition?

**Review by eessex — APPROVED** [review]

File: package.json [hunk 25]
```diff
@@ -25,22 +27,25 @@
...
     "publish-assets": "node scripts/uploadToS3.js",
+    "publish-assets:local": "yarn clean && yarn build:assets && yarn publish-assets",
     "relay": "relay-compiler",
     "sentry": "scripts/sentry.sh",
-    "start:dev": "NODE_ENV=development BUILD_CLIENT=true node --inspect --max_old_space_size=3072 -r dotenv/config ./src",
-    "start:prod:local": "NODE_ENV=production node --inspect --max_old_space_size=3072 -r dotenv/config ./server.dist.js",
-    "start:prod": "yarn assets && yarn build:server && NODE_ENV=production yarn start",
     "start": "scripts/start.sh",
+    "start:dev": "concurrently 'yarn start:dev:relay' 'yarn start:dev:server'",
+    "start:dev:server": "NODE_ENV=development BUILD_CLIENT=true node --inspect --max_old_space_size=3072 -r dotenv/config ./src/index.dev.js",
```
Comment (damassi) [inline package.json:36]: Why is this needed? How is it separate from other commands such as `yarn start` (which product devs will use) or `yarn start:dev:server`?

File: package.json [hunk 26]
```diff
@@ -25,22 +27,25 @@
...
+    "start:dev:relay": "NODE_ENV=development yarn relay --watch",
+    "start:prod": "yarn build:assets && yarn build:server && NODE_ENV=production yarn start",
+    "start:prod:local": "NODE_ENV=production node --inspect --max_old_space_size=3072 -r dotenv/config ./server.dist.js",
```
Comment (damassi) [inline package.json:39]: What does "local" mean in this case? Perhaps `debug` might communicate the intent better since we're attaching the inspector?

File: src/index.js (hunk 13 above)
Comment (zephraph) [reply src/index.js:6]: You should also check w/ different browsers just to be safe.

File: src/common-middleware.js [hunk 27]
```diff
@@ -0,0 +1,84 @@
...
+    .sync(`${__dirname}/{public,{desktop,mobile}/**/public}`)
+    .forEach(folder => {
+      app.use(express.static(folder))
+    })
+
+  app.use(favicon(path.resolve(__dirname, "mobile/public/images/favicon.ico")))
+  app.use("/(.well-known/)?apple-app-site-association", siteAssociation)
+}
+
+export default function commonMiddlewareSetup(app) {
```
Comment (damassi) [inline src/common-middleware.js:71]: (blocking) As mentioned in a previous comment, here and elsewhere we should be using named exports

File: src/lib/manifest.ts (hunk 19 above)
Comment (zephraph) [reply src/lib/manifest.ts:11]: seconded! This is super interesting.

File: src/current.js [hunk 28]
```diff
@@ -0,0 +1,88 @@
+import artsyXapp from "@artsy/xapp"
```
Comment (damassi) [inline src/current.js:1]: `current` is a somewhat confusing name. What does it mean? Current as to what? Perhaps a comment can be added to the top of the file.

File: src/index.js (hunk 13 above)
Comment (damassi) [reply src/index.js:6]: @icirellik Can we delete this and see what happens?

File: src/lib/middleware/intercom.ts [hunk 29]
```diff
@@ -2,7 +2,7 @@ import * as crypto from "crypto"
 // @ts-ignore
```
Comment (zephraph) [inline src/lib/middleware/intercom.ts:2]: Not for this PR, but... do we use intercom anymore? I thought it was removed. 🤔 

@joeyAghion?

**Comment by damassi** [comment]
Not blocking, but I don't quite understand why we need so many new `package.json` scripts. Those that relate to novo assets seem necessary, but otherwise devs have gotten by quite well with: 

```
yarn start
yarn start:prod
yarn jest
yarn mocha
```

File: src/lib/manifest.ts (hunk 19 above)
Comment (icirellik) [reply src/lib/manifest.ts:11]: TypeScript supports constructor assignment, public and private, so that you don't need to do the following:

```ts
class TestClass {
  private manifestFileName: string;

  constructor(manifestFileName: string) {
    this.manifestFileName = manifestFileName;
  }
}
```

File: webpack/envs/clientNovoDevelopmentConfig.js (hunk 20 above)
Comment (icirellik) [reply webpack/envs/clientNovoDevelopmentConfig.js:20]: I plan on cleaning up the WebPack config in a follow-up PR. `weback.smart` has been removed from the library due to many complex edge cases as the author put it and as such we should move away from using it and use `webpack.merge` instead.

Comment (icirellik) [reply webpack/envs/clientNovoDevelopmentConfig.js:20]: Duplication is preferable while novo is tweaked independently of force to simplify testing out changes without breaking the other app. A large amount of effort has been made to isolate novo code from force knowing that there will be temporary duplication.

File: webpack/envs/clientNovoProductionConfig.js (hunk 21 above)
Comment (icirellik) [reply webpack/envs/clientNovoProductionConfig.js:20]: See the previous comment about duplication.

File: webpack/index.js (hunk 22 above)
Comment (icirellik) [reply webpack/index.js:46]: See the previous comment about duplication.

File: src/current.js (hunk 28 above)
Comment (icirellik) [reply src/current.js:1]: I agree, naming is hard. What should we call old force is a hard question. A suggestion would be helpful if you have any and would make for an actionable comment.

File: package.json (hunk 24 above)
Comment (icirellik) [reply package.json:37]: I'll respond to all of the package.json comments here, it's hard to follow your thoughts as they are scattered throughout your review.

In short, yes development for devs hasn't changed in any significant way. I expect it to continue evolving over time and ensuring that existing commands continued to work was the easiest change to make without causing immediate and significant workflow issues.

I personally feel like clearly documenting the commands that are used for force in the `package.json` is preferable for several reasons.

1. Maintaining bash scripts is tedious, the shell usage is inconsistent, error handling is inconsistent, they only execute within a yarn/npm subshell, and is not cross-platform. Both npm and yarn have a minimal shell built in that attempts to solve some of the cross-platform issues.
2. Whenever you want to know a current command does, you incur a side trip penalty because you need to locate and process an additional file to find out what is happening. 
3. Take a look at the current `yarn start` script. Can you confidently tell me which command it will execute? I cannot without inspecting the current shell environment.
4. There are several instances where our code changes paths internally and relies on the current `NODE_ENV` to execute as expected. These changes clearly document that.
5. This is what the `package.json` is typically used for.

So yes there are more commands, they are each complete, and a developer can at a glance tell what that command is. Over time I would expect these "new" commands to slowly replace their counterparts in the scripts folder removing the appearance of duplication.

File: webpack/envs/clientNovoDevelopmentConfig.js (hunk 20 above)
Comment (damassi) [reply webpack/envs/clientNovoDevelopmentConfig.js:20]: I understand the need for independence, but in practical terms we're duplicating 99.9% of the file, for both client and server. Is there any difference between novo webpack config and existing webpack config other than entrypoints?

File: src/lib/middleware/assetMiddleware.ts (hunk 15 above)
Comment (icirellik) [reply src/lib/middleware/assetMiddleware.ts:7]: `export default` is scattered throughout the force code base and it used quite heavily in our newer V2 code, which due to the [bunny theory of code](https://humanwhocodes.com/blog/2015/05/14/the-bunny-theory-of-code/) will be duplicated over time.  

In this case, the majority of our middleware use `module.exports` and keeping with consistency using `export defaults` for the few new ones makes sense in this case. If we wish to do away with `export default` let us make a concerted effort to clean up our codebase once and for all.

File: package.json (hunk 24 above)
Comment (damassi) [reply package.json:37]: Makes sense 👍 Can we remove the yarn start relay command in favor of existing relay command which the team is already familiar with?

File: package.json (hunk 26 above)
Comment (damassi) [reply package.json:39]: Thoughts on using `start:prod:debug`?

File: src/current.js (hunk 28 above)
Comment (damassi) [reply src/current.js:1]: There's a question here (what does `current` mean) as well as a suggestion (comment top of file).

File: src/v2/Apps/getAppNovoRoutes.tsx [hunk 30]
```diff
@@ -0,0 +1,23 @@
+import { buildAppRoutes } from "v2/Artsy/Router/buildAppRoutes"
+import { RouteConfig } from "found"
+import { routes as artistRoutes } from "v2/Apps/Artist/routes"
+import { debugRoutes } from "./Debug/debugRoutes"
+import { cloneDeep } from "lodash"
+
+export function getAppNovoRoutes(): RouteConfig[] {
+  return buildAppRoutes([
+    {
+      routes: cloneDeep(artistRoutes).map(route => {
```
Comment (damassi) [inline src/v2/Apps/getAppNovoRoutes.tsx:10]: Why are we using `cloneDeep` here?

File: src/lib/middleware/intercom.ts (hunk 29 above)
Comment (joeyAghion) [reply src/lib/middleware/intercom.ts:2]: It was abandoned but not removed (tracked in [PURCHASE-2222](https://artsyproduct.atlassian.net/browse/PURCHASE-2222)).

File: src/v2/Apps/getAppNovoRoutes.tsx (hunk 30 above)
Comment (icirellik) [reply src/v2/Apps/getAppNovoRoutes.tsx:10]: We are using a mutable data structure to define the routes and this map operation alters them for both novo and force. I plan on refactoring the routing to use classes or something a bit more stable.

Comment (damassi) [reply src/v2/Apps/getAppNovoRoutes.tsx:10]: Wont cloneDeep then dupliciate everything within the `Component` key of the tree? Like everything inside of [artistRoutes](https://github.com/artsy/force/blob/master/src/v2/Apps/getAppRoutes.tsx#L22) and its references?

Comment (damassi) [reply src/v2/Apps/getAppNovoRoutes.tsx:10]: (Just something to keep an eye on)

**Comment by joeyAghion** [comment]
I'm interested in merging this in order to do some testing in staging over hackathon and before the weekend.

There are worthy tasks pending, but [as discussed in stand-up today] I'd rather they be follow-ups than cause this already long-lived branch to extend into another week. Specifically:
* improved file-naming
* consolidating yarn commands
* refactoring middleware
* abiding by Force's export conventions

Hold us to those!

Hear goes nothing...

File: package.json (hunk 26 above)
Comment (icirellik) [reply package.json:39]: Would it be possible to have a linear conversation? I'm struggling with the constant jumping between threads to discuss related topics.

File: src/current.js (hunk 28 above)
Comment (icirellik) [reply src/current.js:1]: Current means whatever we want to call force, I mentioned that in my previous comment. I don't feel like a comment would add anything additional once the file has found a descriptive name.

File: package.json (hunk 26 above)
Comment (icirellik) [reply package.json:39]: I would suggest that we migrate to standard npm/yarn naming conventions long term.

* yarn start
* yarn dev
* yarn test

and then add one-offs as needed using `:` is a visual way to express intent, that a task belongs to a related topic, and/or is a subtask of an overall task.

`yarn jest` and `yarn mocha` appear to me as being purely historical in meaning due to how the CircleCI orbs were previously configured.

I feel like `debug` is a practical change for `local`. If you weren't aware, for very simple changes like this Github has a handy feature built-in that allows you to suggest the change so that it may be committed directly from the PR.

My issues with the scripts has little to do with bash, I ❤️ bash and would write force in it if I could. My issue is more straightforward. You cannot execute the scripts from the command line:

```sh
# errors
./scripts/acceptance.sh src/test/acceptance/*.js
```

Remember the goal in the PR is to enable the next steps in development for novo without any developer needing to change their current workflow. If you would like to consolidate the commands and scripts to clarify naming and usage, I welcome any outside help.

Comment (damassi) [reply package.json:39]: This makes sense 👍

File: src/v2/Apps/getAppNovoRoutes.tsx (hunk 30 above)
Comment (icirellik) [reply src/v2/Apps/getAppNovoRoutes.tsx:10]: This is a fair concern, I didn't think about what all was attached to the route. I expected that it was a simple configuration object, I doubt there would be any issue with deep cloning here aside from potential wasted memory.

**Comment by joeyAghion** [comment]
We observed 500s on `/news` and [failures of acceptance tests to authenticate](https://app.circleci.com/pipelines/github/artsy/integrity/4237/workflows/52fcb06a-f5ee-450b-b4f6-2709534afb63).

Reverting! Hope to see you again soon...
